context-monitoring/src/main/java/de/atb/context/monitoring/monitors/file/AbstractFileSystemMonitor.java

```diff
@@ -0,0 +1,296 @@
+                        StandardWatchEventKinds.ENTRY_CREATE,
+                        StandardWatchEventKinds.ENTRY_MODIFY,
+                        StandardWatchEventKinds.ENTRY_DELETE,
+                        StandardWatchEventKinds.OVERFLOW);
+                    while (true) {
+                        try {
+                            signalledKey = watchService.take();
+                            Long time = System.currentTimeMillis();
+                            handleWatchEvents(signalledKey.pollEvents(), time);
+                            signalledKey.reset();
+                        } catch (InterruptedException ix) {
+                            this.logger
+                                .info("Watch service was interrupted, closing...");
+                            this.logger.debug(ix.getMessage(), ix);
+                            watchService.close();
+                            Thread.currentThread().interrupt();
+                            break;
+                        } catch (ClosedWatchServiceException cwse) {
+                            this.logger
+                                .info("Watch service closed, terminating...");
+                            this.logger.debug(cwse.getMessage(), cwse);
+                            break;
+                        }
+                    }
+                } catch (IOException e) {
+                    this.logger.error(e.getMessage(), e);
+                }
+            }
+        }, "File watch service Thread");
+        this.watchDaemon.start();
+    }
+
+    protected final void handleWatchEvents(final List<WatchEvent<?>> events,
+                                           final Long time) {
+        String watchedPath = String.valueOf(Paths.get(this.pathToMonitor
+            .getAbsolutePath()));
+        String from = null;
+        for (WatchEvent<?> e : events) {
+            Path context = (Path) e.context();
+            String file = watchedPath + java.io.File.separator
+                + context.toString();
+            InterpreterConfiguration setting = this.interpreter
+                .getConfiguration(file);
+            if (e.kind() == StandardWatchEventKinds.ENTRY_CREATE) {
+                fileCreated(file, time, setting);
+            } else if (e.kind() == StandardWatchEventKinds.ENTRY_MODIFY) {
+                fileModified(file, time, setting);
+            } else if (e.kind() == StandardWatchEventKinds.ENTRY_DELETE) {
+                fileDeleted(file, time, setting);
+            } else {
+                this.logger.debug("Event " + e.kind() + " will be ignored");
+            }
+        }
+    }
+
+    protected abstract void iterateFiles(final File directory);
+
+    protected abstract void handleFile(final String fileName, final Long time,
+                                    final InterpreterConfiguration setting);
+
+    protected abstract IndexingParser<P> getParser(final InterpreterConfiguration setting);
```

Comment:
This method does not need to be abstract because the implementation is the same in all derived classes.

Reply:
- handleFile is not the same in the derived classes
- getParser has different return values

Reply:
I was only referring to `getParser`, which returns `IndexingParser<P>`, where `<P>` is the type parameter of `AbstractFileSystemMonitor`.

So for `FileTripletSystemMonitor extends AbstractFileSystemMonitor<Triplet<File, File, File>>` `getParser` will return a `IndexingParser<Triplet<File, File, File>>`, which is what is needed.

Reply:
[x] done